oeps/oep-0012.rst

```diff
@@ -0,0 +1,475 @@
+It would be preferable for a course tab to be able to register a pluggable user
+interface that will be presented as a tab, and for the implementor not to need
+to know how to implement the full page.
+
+There has been a long standing request to support rendering XBlocks as course
+tabs (`Allow xblocks to be shown as course tabs`_), but there are conceptual
+issues with this approach:
+
+* XBlocks must be configured and saved into the modulestore which doesn't make
+  much sense for course tabs that don't have their own state
+
+* Replacing a built-in tab such as "Course" would require a data migration for
+  every course if it was saved as an XBlock
+
+* It seems confusing to make an XBlock that should never be added into the
+  courseware, but which only exists to be shown as a tab
+
+It would seem better to allow a feature library to register a course tab plugin
+which in turn renders using a component. This allows the feature to provide a
+custom UI without having to add custom blocks to each course, and if the feature
+is removed then the tab will seamlessly disappear without having to remove the
+blocks again.
+
+Use Case 2: Discussions
+-----------------------
+
+Open edX Discussions is a complex feature that provides a variety of
+functionality:
+
+* An inline discussion XBlock for in-context discussions within the courseware
+
+* The "Discussion" tab (aka the discussion board) for course-wide discussions
+
+* A learner profile page that shows all of the discussion posts that the
+  learner has created or commented upon.
+
+* A discussion that is shown within the "Teams" tab for each course team
+
+* In the future, there are thoughts of having program-wide discussions,
+  enterprise-wide discussions, and global discussions with a social group
+
+* There is also the concept of changing course announcements to be a new type of
+  discussion post
+
+In an ideal world, Open edx Discussions would be a separately-installable
+feature that would provide plugins for all of the functionality listed above. An
+Open edX operator could choose not to install edX discussions and choose to
+install different collaboration features instead.
+
+Currently the discussion board is a Django view built into edx-platform, but it
+should instead be made into a plugin. One option would be to create a new
+CourseDiscussionXBlock to display the "Course" tab, but as described above there
+are downsides to forcing this to be an XBlock. It would be cleaner if the edX
+discussion library provided a pluggable user interface that can be dropped into
+each of the contexts in which it is needed. The component would be provided
+with the context for authorization and to customize itself appropriately.
+
+An interesting question is how can such components could share course-scoped
+settings that the instructor specified in Studio. See `OEP-6 Context-scoped XBlock
+Fields`_ for the design of context-scoped fields in XBlock.
```

Comment:
@bradenmacdonald We've had some very interesting discussions about this. The challenge with using a regular Django model is that the settings don't live with the course, so they get lost in an export/import round-trip. If that isn't an issue, then absolutely a Django model is the right choice. If they need to live in the course, then either we need something like context-scoped fields, or we need to provide an extension to Django models to declare how they can be persisted into the OLX.